### PR TITLE
Enable use of pytest and code coverage analysis

### PR DESCRIPTION
**Pull request by berquist**
Issue: https://github.com/cclib/cclib/issues/455

For this to work, it also solves https://github.com/cclib/cclib/issues/367. Everything should be a module now.

Currently Travis is not running this, because the output style isn't great. pytest as a drop-in replacement works, but our tests (in particular regressions) are not in the "right" format. Maybe pytest is only useful for us to run while developing?

- [x] Can I remove all the `if __name__ == '__main__'` calls in the source? Maintaining that is difficult and I don't think they're ever run.

### CODE REVIEW

File: src/cclib/bridge/cclib2biopython.py [hunk 1]
```diff
@@ -12,7 +12,7 @@
     # Fail silently for now.
     pass
 
-from cclib.parser.utils import PeriodicTable
+from ..parser.utils import PeriodicTable
```
Comment (langner) [inline src/cclib/bridge/cclib2biopython.py]: I think with this we should follow PEP8. Absolute imports are recommended, as they are usually more readable and tend to be better behaved: https://www.python.org/dev/peps/pep-0008/#imports

**Comment by langner** [comment]
Whatever we do, I think it would be better to stick to one test framework.

File: src/cclib/bridge/cclib2biopython.py (hunk 1 above)
Comment (berquist) [reply src/cclib/bridge/cclib2biopython.py]: Ok, I made all the changes for `src`, but I'm not sure what to do about imports within the test directory for relative calls.

**Comment by berquist** [comment]
> Whatever we do, I think it would be better to stick to one test framework.

In that case I'll set Travis to use it for now.

**Comment by langner** [comment]
@ATenderholt do you have any problems with using pytest throughout?

**Comment by berquist** [comment]
1. Test suite takes 7 minutes for 2.7 and 11 minutes for 3.6 to run rather than 4 minutes. I'm not sure if this is because coverage is enabled, or because of pytest.

2. Pytest is 2.7 and 3.4+, so it won't work for 3.2.

**Comment by berquist** [comment]
On my desktop, first time is for unit tests, second is for regressions.

* test coverage on: 35.77 s, 147.96 s, 3 min total
* test coverage off: 12.31 s, 48.63 s, 1 min total

So that's where the time difference comes from.

**Comment by ATenderholt** [comment]
I think switching to pytest is a good idea. Seems easier to write and maintain tests, and the code coverage is a nice benefit.

**Comment by langner** [comment]
All right, then it's decided. Eric, I think Travis has an installation issue with pytest for Python 3.2?

**Comment by berquist** [comment]
Yes, 3.4 is the minimum supported 3.x version. Can I bump the Travis version? I have some computers with access to Ubuntu 14.04, which is no longer supported, and they have 3.4.3.

**Comment by langner** [comment]
I see. I'm somewhat nervous about dropping support for a Python version we already supported.

**Comment by berquist** [comment]
I fixed Python 3.2 by installing ancient versions of pytest + dependencies. It may break if we start using pytest-only features on our tests.

1. Something is wrong with coverage checking. I'm not sure if checking the two trees (see Travis log) is messing up the statistics, because I thought testing regressions would make it go up.

2. Right now output from `test_all` is being silenced. I'm not sure what it will look like when a test inside of it fails.

File: test/method/testnuclear.py [hunk 2]
```diff
@@ -17,12 +17,14 @@
 
 import numpy
 
-sys.path.append("..")
-from test_data import getdatafile
 from cclib.method import Nuclear
 from cclib.parser import QChem
 from cclib.parser import utils
 
+sys.path.append("..")
```
Comment (dgasmith) [inline test/method/testnuclear.py]: To be safe, shouldn't this be `sys.path.insert(1, "..")`.

**Comment by dgasmith** [comment]
Big fan of `pytest` overall, after getting used to it I use it for all of my projects.

File: conftest.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
+import sys
+
+
+version_major = sys.version_info.major
+version_minor = sys.version_info.minor
+
+# Paths than should be ignored for all Python versions.
```
Comment (langner) [inline conftest.py]: than->that

**Comment by berquist** [comment]
Ok, I’ll turn doctests into unit tests and remove all doctests from the
codebase.

As far as I can tell, pytest is dependent on solidifying the import
situation. I am also going to try and do the path prepend  Daniel suggested.

I’ll break it apart right now, just need some coffee...

**Comment by berquist** [comment]
For those who see this in the future and wonder why it needs to be `sys.path.insert(1, '..')` and not `sys.path.insert(0, '..')`: https://stackoverflow.com/q/10095037

Short answer: it will mess with using virtualenvs.

**Comment by langner** [comment]
> For those who see this in the future and wonder why it needs to be sys.path.insert(1, '..') and not sys.path.insert(0, '..'): https://stackoverflow.com/q/10095037
> 
> Short answer: it will mess with using virtualenvs.

Could you add a comment to this effect near the code?

**Comment by berquist** [comment]
The new PR is https://github.com/cclib/cclib/pull/491.

**Pull request closed without merging**